### PR TITLE
[cleanup] Cleaned up crystal damage calculations

### PR DESCRIPTION
**Pull request by nopjmp**
CombatSettings was not staying on the main thread where the tick was happening. If the coroutines on the defaultScope thread pool accessed the entityLists, they could be in a different state and have concurrent modification going on and not geting the most accurate information for that tick.

The alternative solution that I could add here is to grab the entityList and make a deep copy of the information that we are going to use in the checks before passing it in.

These run pretty fast as is in most situations; however, I think this is a good approach for the edge cases where a bigger PVP battle is happening and more entities are at play.

Additionally, move the Tripe<> tuple to a named data class for calculations and add a comparator that sorts by target damage and self damage, distance is taken account in the placement loop.

### CODE REVIEW

**Review by l1ving — APPROVED** [review]

File: src/main/kotlin/org/kamiblue/client/module/modules/combat/CombatSetting.kt [hunk 1]
```diff
@@ -131,9 +130,10 @@ internal object CombatSetting : Module(
...
 
         safeListener<TickEvent.ClientTickEvent>(5000) {
-            for ((function, future) in jobMap) {
-                if (future.isActiveOrFalse) continue // Skip if the previous thread isn't done
-                jobMap[function] = defaultScope.launch { function(this@safeListener) }
+            runBlocking {
+                for (function in taskList) {
+                    defaultScope.launch { function(this@safeListener) }
+                }
             }
```
Comment (Xiaro) [inline src/main/kotlin/org/kamiblue/client/module/modules/combat/CombatSetting.kt:137]: You don't need runBlocking if you are launching it on the defaultScope

File: src/main/kotlin/org/kamiblue/client/module/modules/combat/CombatSetting.kt [hunk 2]
```diff
@@ -96,10 +95,10 @@ internal object CombatSetting : Module(
...
     private val resumeTimer = TickTimer(TimeUnit.SECONDS)
-    private val jobMap = hashMapOf<(SafeClientEvent) -> Unit, Job?>(
-        { it: SafeClientEvent -> it.updateTarget() } to null,
-        { it: SafeClientEvent -> it.updatePlacingList() } to null,
-        { it: SafeClientEvent -> it.updateCrystalList() } to null
+    private val taskList = arrayListOf<(SafeClientEvent) -> Unit>(
+        { it.updateTarget() },
+        { it.updatePlacingList() },
+        { it.updateCrystalList() }
     )
```
Comment (Xiaro) [inline src/main/kotlin/org/kamiblue/client/module/modules/combat/CombatSetting.kt:102]: This is for preventing starting new calculations if the previous one wasn't done yet

File: src/main/kotlin/org/kamiblue/client/module/modules/combat/CombatSetting.kt [hunk 3]
```diff
@@ -161,27 +161,27 @@ internal object CombatSetting : Module(
...
         for (pos in getPlacePos(target, player, 8f)) {
             val dist = eyePos.distanceTo(pos.toVec3dCenter(0.0, 0.5, 0.0))
             val damage = target?.let { calcCrystalDamage(pos, it, prediction?.first, prediction?.second) } ?: 0.0f
             val selfDamage = calcCrystalDamage(pos, player)
-            cacheList.add(Pair(pos, Triple(damage, selfDamage, dist)))
+            cacheList.add(Pair(pos, CombatManager.CombatCalculation(damage, selfDamage, dist)))
         }
 
-        CombatManager.placeMap = LinkedHashMap<BlockPos, Triple<Float, Float, Double>>(cacheList.size).apply {
-            putAll(cacheList.sortedByDescending { triple -> triple.second.first })
```
Comment (Xiaro) [inline src/main/kotlin/org/kamiblue/client/module/modules/combat/CombatSetting.kt:176]: sortedWith already takes Comparator as parameter

**Review by Xiaro — CHANGES_REQUESTED** [review]

**Comment by Xiaro** [comment]
Blocking the main thread is not necessary and it impacts performance. Blocks related stuff in Minecraft is pretty much thread safe. It also make a copy on entityList and armor list for thread safety. Lastly the calculation doesn't need to be accurate but rather to be fast. So the threading related change is unnecessary but the data class for damage calculation result is good.

**Comment by nopjmp** [comment]
> Blocks related stuff in Minecraft is pretty much thread safe. It also make a copy of entityList and armor list for thread safety.

It's not thread safe completely. I had forgotten that reference object replacement and fetching is thread safe per the Java spec; however, double values are only thread safe on 64bit platforms due to the word size being 64bit and they happen to be thread safe. We unfortunately, have to think about these conditions due to Vec3d and other primitives being double/long instead of float/int.

This means we could be looking at a variable half way through a write on another thread on a 32bit operating system. This may or may not be a problem since we are dealing with entities that are within the same fraction or close, but it can result in jittering around numbers. This is only the case because Entities have mutable state in their coords and we are only taking a copy of the list which copies over the direct reference to the objecft. BlockPos is immutable (outside of MutableBlockPos).

This all just feels "code smelly" to me and seems like a lot of additional care to the map usage. I'm still trying to think of a way to better utilize kotlin.coroutinex around this, possibly using the [StateFlow](https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/-state-flow/) which would cement the contract in the interface and possibly allow these entity fetching tasks to just be a background thread to do the work.

File: src/main/kotlin/org/kamiblue/client/module/modules/combat/CombatSetting.kt (hunk 2 above)
Comment (nopjmp) [reply src/main/kotlin/org/kamiblue/client/module/modules/combat/CombatSetting.kt:102]: I understand that, but this is in support of the behavioral changes I made.

File: src/main/kotlin/org/kamiblue/client/module/modules/combat/CombatSetting.kt (hunk 1 above)
Comment (nopjmp) [reply src/main/kotlin/org/kamiblue/client/module/modules/combat/CombatSetting.kt:137]: I explained in the comment in the PR about my reasoning behind changing this from background coroutines to blocking on each update tick.

Alternative also present as a way forward and I can revert this behavior and implement that variation.

File: src/main/kotlin/org/kamiblue/client/module/modules/combat/CombatSetting.kt (hunk 3 above)
Comment (nopjmp) [reply src/main/kotlin/org/kamiblue/client/module/modules/combat/CombatSetting.kt:176]: sortedWith takes a Comparator yes, but it's not compatible with the CombatCalculations comparator which is why it's 

compareBy(comparator) { it.second } which unwraps the key part of the pair to allow for the comparator usage.

**Comment by Xiaro** [comment]
> > Blocks related stuff in Minecraft is pretty much thread safe. It also make a copy of entityList and armor list for thread safety.
> 
> It's not thread safe completely. I had forgotten that reference object replacement and fetching is thread safe per the Java spec; however, double values are only thread safe on 64bit platforms due to the word size being 64bit and they happen to be thread safe. We unfortunately, have to think about these conditions due to Vec3d and other primitives being double/long instead of float/int.
> 
> This means we could be looking at a variable half way through a write on another thread on a 32bit operating system. This may or may not be a problem since we are dealing with entities that are within the same fraction or close, but it can result in jittering around numbers. This is only the case because Entities have mutable state in their coords and we are only taking a copy of the list which copies over the direct reference to the objecft. BlockPos is immutable (outside of MutableBlockPos).

That is very edge case and most(over 95%) players uses 64 bit operation system nowadays. It also make a copy of the target's coordinate for calculations already. I still don't see an advantage of sacrificing performance for a bit more accuracy.

**Comment by nopjmp** [comment]
> That is very edge case and most(over 95%) players uses 64 bit operation system nowadays. It also make a copy of the target's coordinate for calculations already. I still don't see an advantage of sacrificing performance for a bit more accuracy.

I still think targeting should happen first before the other checks just to help change targets more accurately.

I'll remove the coroutine changes for now. I still believe we should define the relationship better using modern coroutine features that are coming along in Kotlin.

I only encountered a single "bug" where the server kicked me for sending break / attack packets to an invalid entity id. I'm not sure how we would combat this from happening without staying in sync with the ticks which we could have as an option and switch between blocking and non-blocking with non-blocking being the default.

**Review by Xiaro — APPROVED** [review]

**Review by l1ving — APPROVED** [review]

**Comment by l1ving** [comment]
Thanks!

**Pull request merged**